scio-elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -204,102 +221,168 @@ public void processElement(ProcessContext c) throws Exception {
         this.clientSupplier = new ClientSupplier(clusterName, servers);
         this.toActionRequests = toActionRequests;
         this.error = error;
       }
+
       @ProcessElement
       public void processElement(ProcessContext c) throws Exception {
-        final List<Iterable<ActionRequest<?>>> actionRequests =
+        final List<Iterable<ActionRequest>> actionRequests =
             StreamSupport.stream(c.element().getValue().spliterator(), false)
                 .map(toActionRequests::apply)
                 .collect(Collectors.toList());
-
         // Elasticsearch throws ActionRequestValidationException if bulk request is empty,
         // so do nothing if number of actions is zero.
         if (actionRequests.isEmpty()) {
           LOG.info("ElasticsearchWriter: no requests to send");
           return;
         }
 
-        final BulkRequest bulkRequest = new BulkRequest().add(Iterables.concat(actionRequests));
-        final BulkResponse bulkItemResponse = clientSupplier.get().bulk(bulkRequest).get();
+        final Client client = clientSupplier.get();
+        final BulkRequestBuilder bulkRequestBuilder = client.prepareBulk();
+        Iterables.concat(actionRequests)
+            .forEach(x -> {
+              try {
+                addBulkRequest(client, bulkRequestBuilder, x);
+              } catch (IOException e) {
+                throw new UncheckedIOException(e);
+              }
+            });
+        final BulkResponse bulkItemResponse = bulkRequestBuilder.get();
+
         if (bulkItemResponse.hasFailures()) {
           error.accept(new BulkExecutionException(bulkItemResponse));
         }
       }
     }
 
-    private static class ClientSupplier implements Supplier<Client>, Serializable {
-      private final AtomicReference<Client> CLIENT = new AtomicReference<>();
-      private final String clusterName;
-      private final InetSocketAddress[] addresses;
-
-      public ClientSupplier(final String clusterName, final InetSocketAddress[] addresses) {
-        this.clusterName = clusterName;
-        this.addresses = addresses;
-      }
-      @Override
-      public Client get() {
-        if (CLIENT.get() == null) {
-          synchronized (CLIENT) {
-            if (CLIENT.get() == null) {
-              CLIENT.set(create(clusterName, addresses));
-            }
+    /**
+     * Accepts a Client, a BulkRequestBuilder, and an ActionRequest.
+     * Properly typecasts & prepares an ActionRequest. Adds it to the BulkRequestBuilder.
+     * Return type is void.
+     *
+     * @param client        provides the ability to prepare requests for bulk execution
+     * @param bulkRequest   maintains an ordered list of ActionRequests to later be executed in bulk
+     * @param actionRequest the ActionRequest to be added to the bulkRequest
+     */
+    private static void addBulkRequest(Client client, BulkRequestBuilder bulkRequest,
```

Comment:
ES lib already does it https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java#L109